fili-core/src/main/java/com/yahoo/bard/webservice/web/responseprocessors/EtagCacheResponseProcessor.java

```diff
@@ -0,0 +1,136 @@
+
+import javax.validation.constraints.NotNull;
+import javax.ws.rs.core.Response.Status;
+
+/**
+ * A response processor which caches the results if appropriate after completing a query according to etag value.
+ */
+public class EtagCacheResponseProcessor implements FullResponseProcessor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EtagCacheResponseProcessor.class);
+
+    private final ObjectMapper mapper;
+    private final ObjectWriter writer;
+    private final ResponseProcessor next;
+    private final QuerySigningService<Long> querySigningService;
+    private final @NotNull TupleDataCache<String, Long, String> dataCache;
+
+
+    /**
+     * Constructor.
+     *
+     * @param next  Next ResponseProcessor in the chain
+     * @param dataCache  The cache into which to write a cache entry
+     * @param querySigningService  Service to use for signing the queries in the cache key with their metadata
+     * @param mapper  An object mapper to use for processing Json
+     */
+    public EtagCacheResponseProcessor(
+            ResponseProcessor next,
+            TupleDataCache<String, Long, String> dataCache,
+            QuerySigningService<Long> querySigningService,
+            ObjectMapper mapper
+    ) {
+        this.next = next;
+        this.dataCache = dataCache;
+        this.querySigningService = querySigningService;
+        this.mapper = mapper;
+        this.writer = mapper.writer();
+    }
+
+    @Override
+    public ResponseContext getResponseContext() {
+        return next.getResponseContext();
+    }
+
+    @Override
+    public FailureCallback getFailureCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getFailureCallback(druidQuery);
+    }
+
+    @Override
+    public HttpErrorCallback getErrorCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getErrorCallback(druidQuery);
+    }
+
+    @Override
+    public void processResponse(JsonNode json, DruidAggregationQuery<?> druidQuery, LoggingContext metadata) {
+        // make sure JSON response comes with status code
+        if (!json.has(DruidJsonResponseContentKeys.STATUS_CODE.getName())) {
+            String message = ErrorMessageFormat.STATUS_CODE_MISSING_FROM_RESPONSE.format();
+            LOG.error(message);
+            getErrorCallback(druidQuery).dispatch(
+                    Status.INTERNAL_SERVER_ERROR.getStatusCode(),
+                    "The server encountered an unexpected condition which prevented it from fulfilling the request.",
+                    message
+            );
+        }
+
+        int statusCode = json.get(DruidJsonResponseContentKeys.STATUS_CODE.getName()).asInt();
+        if (statusCode == Status.NOT_MODIFIED.getStatusCode()) {
+            try {
+                ((ObjectNode) json).set(
+                        DruidJsonResponseContentKeys.RESPONSE.getName(),
+                        mapper.readTree(dataCache.getDataValue(DruidJsonResponseContentKeys.RESPONSE.getName()))
```

Comment:
It might be slightly more efficient to keep this around rather than reserialize on cache hit.  But I don't think it matters much.